lib/services/unified_resource.go

```diff
@@ -0,0 +1,675 @@
+	re := c.getRange(ctx, startKey, endKey, limit)
+	if len(re.Items) == backend.DefaultRangeLimit {
+		c.log.Warnf("Range query hit backend limit. (this is a bug!) startKey=%q,limit=%d", startKey, backend.DefaultRangeLimit)
+	}
+	return &re, nil
+}
+
+// CompareAndSwap compares item with existing item and replaces it with replaceWith item
+func (c *unifiedResourceCache) CompareAndSwap(ctx context.Context, expected Item, replaceWith Item) error {
+	if len(expected.Key) == 0 {
+		return trace.BadParameter("missing parameter Key")
+	}
+	if len(replaceWith.Key) == 0 {
+		return trace.BadParameter("missing parameter Key")
+	}
+	if !bytes.Equal(expected.Key, replaceWith.Key) {
+		return trace.BadParameter("expected and replaceWith keys should match")
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	i, found := c.tree.Get(&btreeItem{Item: expected})
+	if !found {
+		return trace.CompareFailed("key %q is not found", string(expected.Key))
+	}
+	existingItem := i.Item
+	if existingItem.Value != expected.Value {
+		return trace.CompareFailed("current value does not match expected for %v", string(expected.Key))
+	}
+	event := Event{
+		Type: types.OpPut,
+		Item: replaceWith,
+	}
+	c.processEvent(event)
+	return nil
+}
+
+type getResult struct {
+	Items []Item
+}
+
+func (c *unifiedResourceCache) getRange(ctx context.Context, startKey, endKey []byte, limit int) getResult {
+	var res getResult
+	c.tree.AscendRange(&btreeItem{Item: Item{Key: startKey}}, &btreeItem{Item: Item{Key: endKey}}, func(item *btreeItem) bool {
+		res.Items = append(res.Items, item.Item)
+		if limit > 0 && len(res.Items) >= limit {
+			return false
+		}
+		return true
+	})
+	return res
+}
+
+func (c *unifiedResourceCache) processEvent(event Event) {
+	switch event.Type {
+	case types.OpPut:
+		item := &btreeItem{Item: event.Item, index: -1}
+		c.tree.ReplaceOrInsert(item)
+	case types.OpDelete:
+		item, found := c.tree.Get(&btreeItem{Item: event.Item})
+		if !found {
+			return
+		}
+		c.tree.Delete(item)
+	default:
+		// skip unsupported record
+	}
+}
+
+// UnifiedResourceWatcherConfig is a UnifiedResourceWatcher configuration.
+type UnifiedResourceWatcherConfig struct {
+	ResourceWatcherConfig
+	NodesGetter
+	DatabaseServersGetter
+	AppServersGetter
+	WindowsDesktopGetter
+	KubernetesClusterGetter
+	SAMLIdpServiceProviderGetter
+}
+
+// UnifiedResourceWatcher is built on top of resourceWatcher to monitor additions
+// and deletions to resources displayable in the UI such as Nodes, Dbs, Apps, etc.
+type UnifiedResourceWatcher struct {
+	*resourceWatcher
+	*unifiedResourceCollector
+}
+
+// Close closes the underlying resource watcher
+func (u *UnifiedResourceWatcher) Close() error {
+	u.resourceWatcher.Close()
+	return u.unifiedResourceCollector.Close()
+}
+
+// GetUnifiedResources returns a list of all resources stored in the current unifiedResourceCollector tree
+func (u *UnifiedResourceWatcher) GetUnifiedResources(ctx context.Context) ([]types.ResourceWithLabels, error) {
```

Comment:
The primary motivation of creating a unified resource watcher was to ensure that we _didn't_ need to load the entire resource set into memory simultaneously for each request, and that we'd instead be able to paginate and pre-filter.  Loading the entire resource set into memory per-request is a recipe for OOM errors (not hypothetical, we see this kind of thing happen in real teleport deployments).  This should be a paginated API.  Nothing wrong with having a "get all" method for tests, but it should be private to ensure folks don't accidentally use it to handle actual requests.